File: tests/node/domain/core-kp-routing.test.ts
import { describe, it, expect } from "vitest";
import {
  CORE_TO_KP_MAPPING,
  CONDITIONAL_CORE_KP,
  getActiveRoutes,
} from "../../../src/domain/core-kp-routing.js";
import { aDomainTestConfig } from "../../fixtures/project-config.fixture.js";

describe("CORE_TO_KP_MAPPING", () => {
  it("contains_12_staticRoutes", () => {
    expect(CORE_TO_KP_MAPPING).toHaveLength(12);
  });

  it("firstRoute_isCleanCode", () => {
    const first = CORE_TO_KP_MAPPING[0]!;
    expect(first.sourceFile).toBe("01-clean-code.md");
    expect(first.kpName).toBe("coding-standards");
    expect(first.destFile).toBe("clean-code.md");
  });

  it("lastRoute_isRefactoringGuidelines", () => {
    const last = CORE_TO_KP_MAPPING[11]!;
    expect(last.sourceFile).toBe("14-refactoring-guidelines.md");
    expect(last.kpName).toBe("coding-standards");
    expect(last.destFile).toBe("refactoring-guidelines.md");
  });

  it("allDestFiles_areUnique_noDuplicates", () => {
    const destFiles = CORE_TO_KP_MAPPING.map((r) => r.destFile);
    const uniqueDestFiles = new Set(destFiles);
    expect(uniqueDestFiles.size).toBe(destFiles.length);
  });
});

describe("CONDITIONAL_CORE_KP", () => {
  it("contains_1_conditionalRoute", () => {
    expect(CONDITIONAL_CORE_KP).toHaveLength(1);
  });

  it("conditionalRoute_isCloudNative", () => {
    const route = CONDITIONAL_CORE_KP[0]!;
    expect(route.sourceFile).toBe("12-cloud-native-principles.md");
    expect(route.kpName).toBe("infrastructure");
    expect(route.conditionField).toBe("architecture_style");
    expect(route.conditionExclude).toBe("library");
  });
});

describe("getActiveRoutes", () => {
  it("microservice_includes13Routes", () => {
    const config = aDomainTestConfig({ style: "microservice" });
    const routes = getActiveRoutes(config);
    expect(routes).toHaveLength(13);
  });

  it("library_excludesCloudNative_returns12Routes", () => {
    const config = aDomainTestConfig({ style: "library" });
    const routes = getActiveRoutes(config);
    expect(routes).toHaveLength(12);
    const sourceFiles = routes.map((r) => r.sourceFile);
    expect(sourceFiles).not.toContain("12-cloud-native-principles.md");
  });

  it("monolith_includesCloudNative", () => {
    const config = aDomainTestConfig({ style: "monolith" });
    const routes = getActiveRoutes(config);
    expect(routes).toHaveLength(13);
    const sourceFiles = routes.map((r) => r.sourceFile);
    expect(sourceFiles).toContain("12-cloud-native-principles.md");
  });

  it("allRoutes_haveAllFields", () => {
    const config = aDomainTestConfig({ style: "microservice" });
    const routes = getActiveRoutes(config);
    for (const route of routes) {
      expect(route.sourceFile).toBeTruthy();
      expect(route.kpName).toBeTruthy();
      expect(route.destFile).toBeTruthy();
    }
  });
});

File: java/src/test/java/dev/iadev/domain/stack/CoreKpRoutingTest.java
package dev.iadev.domain.stack;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayName("CoreKpRouting")
class CoreKpRoutingTest {

    @Nested
    @DisplayName("CORE_TO_KP_MAPPING")
    class CoreToKpMappingTests {

        @Test
        @DisplayName("contains exactly 12 static routes")
        void coreMapping_size_twelve() {
            assertThat(CoreKpRouting.CORE_TO_KP_MAPPING).hasSize(12);
        }

        @Test
        @DisplayName("first route maps clean-code to coding-standards")
        void coreMapping_first_cleanCode() {
            var first = CoreKpRouting.CORE_TO_KP_MAPPING.get(0);
            assertThat(first.sourceFile())
                    .isEqualTo("01-clean-code.md");
            assertThat(first.kpName())
                    .isEqualTo("coding-standards");
            assertThat(first.destFile())
                    .isEqualTo("clean-code.md");
        }

        @Test
        @DisplayName("last route maps refactoring to coding-standards")
        void coreMapping_last_refactoring() {
            var last = CoreKpRouting.CORE_TO_KP_MAPPING
                    .get(CoreKpRouting.CORE_TO_KP_MAPPING.size() - 1);
            assertThat(last.sourceFile())
                    .isEqualTo("14-refactoring-guidelines.md");
            assertThat(last.kpName())
                    .isEqualTo("coding-standards");
            assertThat(last.destFile())
                    .isEqualTo("refactoring-guidelines.md");
        }

        @Test
        @DisplayName("list is immutable")
        void coreMapping_whenCalled_immutable() {
            assertThat(CoreKpRouting.CORE_TO_KP_MAPPING)
                    .isUnmodifiable();
        }
    }

    @Nested
    @DisplayName("CONDITIONAL_CORE_KP")
    class ConditionalCoreKpTests {

        @Test
        @DisplayName("contains exactly 1 conditional route")
        void conditionalKp_size_one() {
            assertThat(CoreKpRouting.CONDITIONAL_CORE_KP).hasSize(1);
        }

        @Test
        @DisplayName("conditional route excludes library style")
        void conditionalKp_whenCalled_excludesLibrary() {
            var route = CoreKpRouting.CONDITIONAL_CORE_KP.get(0);
            assertThat(route.sourceFile())
                    .isEqualTo("12-cloud-native-principles.md");
            assertThat(route.conditionField())
                    .isEqualTo("architecture_style");
            assertThat(route.conditionExclude())
                    .isEqualTo("library");
        }
    }

    @Nested
    @DisplayName("getActiveRoutes()")
    class GetActiveRoutesTests {

        @Test
        @DisplayName("microservice includes cloud-native route (13 total)")
        void getActiveRoutes_microservice_includesCloudNative() {
            var config = new TestProjectConfigBuilder()
                    .architectureStyle("microservice")
                    .build();

            var routes = CoreKpRouting.getActiveRoutes(config);

            assertThat(routes).hasSize(13);
            assertThat(routes.stream()
                    .anyMatch(r -> "12-cloud-native-principles.md"
                            .equals(r.sourceFile())))
                    .isTrue();
        }

        @Test
        @DisplayName("library excludes cloud-native route (12 total)")
        void getActiveRoutes_library_excludesCloudNative() {
            var config = new TestProjectConfigBuilder()
                    .architectureStyle("library")
                    .build();

            var routes = CoreKpRouting.getActiveRoutes(config);

            assertThat(routes).hasSize(12);
            assertThat(routes.stream()
                    .noneMatch(r -> "12-cloud-native-principles.md"
                            .equals(r.sourceFile())))
                    .isTrue();
        }

        @Test
        @DisplayName("monolith includes cloud-native route")
        void getActiveRoutes_monolith_includesCloudNative() {
            var config = new TestProjectConfigBuilder()
                    .architectureStyle("monolith")
                    .build();

            var routes = CoreKpRouting.getActiveRoutes(config);

            assertThat(routes).hasSize(13);
        }
    }

    @Nested
    @DisplayName("resolveConditionValue()")
    class ResolveConditionValueTests {

        @Test
        @DisplayName("architecture_style returns config style")
        void resolveCondition_architectureStyle_returnsStyle() {
            var config = new TestProjectConfigBuilder()
                    .architectureStyle("microservice")
                    .build();

            var value = CoreKpRouting.resolveConditionValue(
                    config, "architecture_style");

            assertThat(value).isEqualTo("microservice");
        }

        @Test
        @DisplayName("unknown field returns empty string")
        void resolveCondition_unknownField_empty() {
            var config = new TestProjectConfigBuilder().build();

            var value = CoreKpRouting.resolveConditionValue(
                    config, "unknown_field");

            assertThat(value).isEmpty();
        }
    }
}
